### ISSUE TITLE
Autorun composer install

### ISSUE THREAD

**Issue by SamMousa** [ngineered/nginx-php-fpm#95]
Doesn't it make sense to run composer install on start if a composer.lock is found?

**Comment by richarvey** [comment]
would this need to be run if the site was updated also?

**Comment by SamMousa** [comment]
Yes!

**Comment by richarvey** [comment]
Ok where would the composer file live? Would you always include in the root? Also would it be the lock file we look for or a json file?

**Comment by SamMousa** [comment]
A lock file definitely and I think it must be in the root (not sure)

**Comment by richarvey** [comment]
@SamMousa are you sure about the lock file i thought that was auto created when you ran composer:

https://daylerees.com/the-composer-lock-file/

I may be wrong here so any help greatly appreciated. From reading that on a git pull I'd run composer install if there was a composer.json file in the root directory.

**Comment by SamMousa** [comment]
Yes it is created and contains the exact commit hash of each dependency.
For non libraries it is committed to version control (to guarantee a test
environment exactly the same as the production environment).

See
https://getcomposer.org/doc/01-basic-usage.md#composer-lock-the-lock-file

**Comment by richarvey** [comment]
We'll spin up a new branch asap with this in to test.

**Comment by richarvey** [comment]
Hi @SamMousa I've committed some code that "should" run install when you first launch the container and if you do a pull to update the code it should run an update.

Can you check this functionality behaves as expected.

Cheers Ric

**Comment by SamMousa** [comment]
Will do next week.

**Comment by SamMousa** [comment]
@richarvey Just had a chance to look at the implementation and it doesn't work for my case.

The convention is to put `composer.lock` is in the project root, not in the webroot. My project(s) use a `public` folder as the webroot, the code you've added searches for composer.lock in this webroot directory; instead it should be searched for in the project root. I'll add some comments to the commit as well.

**Comment by SamMousa** [comment]
Also note, that there is technically no need to check for the existence of a composer lock file; if you just ru composer install you'll get this in case it can't run:
````
$ composer install
Composer could not find a composer.json file in /tmp
To initialize a project, please create a composer.json file as described in the https://getcomposer.org/ "Getting Started" section
````

**Comment by richarvey** [comment]
Right I think i understand now. Let me add some variables and see if we can get this working.

**Comment by SamMousa** [comment]
@richarvey Also add `--no-dev`, since dev dependencies shouldn't be installed on production environment (alternatively use a variable for this)

**Comment by SKoschnicke** [comment]
I don't think this is working is intended. The startup script checks for a composer.lock file in the *webroot* and executes the composer in the *project root* then.

If webroot is not the same as project root, either the check for the composer.lock file does not work (because composer.lock is located in the project root) or the composer command does not find the composer.lock file (because the command is executed in the project root and the composer.lock is located in the webroot).

Also, the composer.phar binary needs to reside in the repository (in the project root). I don't think that its goot practice to put this binary artifact into the source repository. The image itself should come with a composer binary.

I'm happy to help with this issue, if you want. :)

**Comment by SamMousa** [comment]
@SKoschnicke I think the run command is just wrong. Composer is already included in the image.

https://github.com/ngineered/nginx-php-fpm/blob/master/Dockerfile#L177

**Comment by SKoschnicke** [comment]
I see. Good to know! The other problems are still valid, though.

**Comment by richarvey** [comment]
@SKoschnicke sorry for the slow reply. I'd love some help with this. I've just merged in a better way to install composer but could you help with fixing the other issues please?

**Comment by SKoschnicke** [comment]
I'm currently ill but I think I can look into it on monday!

**Comment by SKoschnicke** [comment]
In case nobody noticed: I fixed the issue in #130

**Comment by richarvey** [comment]
@SKoschnicke Apologies been really busy will get this done asap
